src/tunnel.c

```diff
@@ -439,44 +443,87 @@ static int tcp_connect(struct tunnel *tunnel)
+		if (bytes_written != strlen(request)) {
+			log_error("write error while talking to proxy: %s\n",
+			          strerror(errno));
 			goto err_connect;
 		}
 
 		// wait for a "200 OK" reply from the proxy,
-		// be careful not to fetch too many characters at once
-		char *response;
-		int bytes_read, curr_pos = 0;
+		// be careful not to fetch too many bytes at once
+		const char *response = NULL;
 
 		memset(&(request), '\0', sizeof(request));
-		do {
-			bytes_read = read(handle, &(request[curr_pos++]), 1);
-			response = strstr(request, "200");
-		} while (
-		        (
-		                // repeat reading until we have got the string "200"
-		                (response == NULL)
-		                // continue reading until we hit the first empty line
-		                || (
-		                        // server uses newlines only
-		                        (strstr(response, "\n\n") == NULL)
-		                        // server uses cr+lf
-		                        && (strstr(response, "\r\n\r\n") == NULL)
-		                        // server uses lf+cr
-		                        && (strstr(response, "\n\r\n\r") == NULL)
-		                )
-		        )
-		        // condition for buffer full or possibly eof
-		        && (curr_pos < sizeof(request) - 1) && (bytes_read > 0)
-		);
-		free(env_proxy); // we have copied the string
+		for (int j = 0; response == NULL; j++) {
+			/*
+			 * Coverity detected a defect:
+			 *  CID 200508: String not null terminated (STRING_NULL)
+			 *
+			 * It is actually a false positive:
+			 * • Function memset() initializes 'request' with '\0'
+			 * • Function read() gets a single char into: request[j]
+			 * • The final '\0' cannot be overwritten because:
+			 *   	j < ARRAY_SIZE(request) - 1
+			 */
+			ssize_t bytes_read = read(handle, &(request[j]), 1);
+			if (bytes_read < 1) {
+				log_error("Proxy response is unexpectedly large and"
+				          " cannot fit in the %d-bytes buffer.\n",
+				          ARRAY_SIZE(request));
+				goto err_proxy_response;
+			}
+
+			// detect "200"
+			const char HTTP_STATUS_200[] = "200";
+			response = strstr(request, HTTP_STATUS_200);
+
+			// detect end-of-line after "200"
+			if (response != NULL) {
+				/*
+				 * RFC2616 states in section 2.2 Basic Rules:
+				 * 	CR     = <US-ASCII CR, carriage return (13)>
+				 * 	LF     = <US-ASCII LF, linefeed (10)>
+				 * 	HTTP/1.1 defines the sequence CR LF as the
+				 * 	end-of-line marker for all protocol elements
+				 * 	except the entity-body (see appendix 19.3
+				 * 	for tolerant applications).
+				 * 		CRLF   = CR LF
+				 *
+				 * RFC2616 states in section 19.3 Tolerant Applications:
+				 * 	The line terminator for message-header fields
+				 * 	is the sequence CRLF. However, we recommend
+				 * 	that applications, when parsing such headers,
+				 * 	recognize a single LF as a line terminator
+				 * 	and ignore the leading CR.
+				 */
+				static const char *HTTP_EOL[] = {
+					"\r\n\r\n",
+					"\n\n"
```

Comment:
what about the "\n\r\n\r" case? The previous code was a bit more tolerant here, but probably this behavior was against specifications anyway...

Reply:
It's indeed against specifications. Therefore I would suggest starting with these two variants and adding `\n\r\n\r`if needed.Your call :-)